optaplanner-core/src/main/java/org/optaplanner/core/api/score/buildin/bendablebigdecimal/BendableBigDecimalScoreHolder.java

```diff
@@ -80,7 +80,7 @@ public void undo() {
 
     public void addSoftConstraintMatch(RuleContext kcontext, final int softLevel, final BigDecimal weight) {
         softScores[softLevel] = softScores[softLevel].add(weight);
-        registerBigDecimalConstraintMatch(kcontext, softLevel, weight, new BigDecimalConstraintUndoListener() {
+        registerBigDecimalConstraintMatch(kcontext, getHardLevelsSize() + softLevel, weight, new BigDecimalConstraintUndoListener() {
```

Comment:
Nice fix - it might even deserve it's own jira as it means that BENDABLE_BIG_DECIMAL score was unreliable, apparently since JUN-2014 according to git log.